lib/pgBackRest/Storage/Filter/CipherBlock.pm

```diff
@@ -0,0 +1,177 @@
+####################################################################################################################################
+# Block Cipher Filter
+####################################################################################################################################
+package pgBackRest::Storage::Filter::CipherBlock;
+use parent 'pgBackRest::Common::Io::Filter';
+
+use strict;
+use warnings FATAL => qw(all);
+use Carp qw(confess);
+use English '-no_match_vars';
+
+use Exporter qw(import);
+    our @EXPORT = qw();
+
+use pgBackRest::Common::Exception;
+use pgBackRest::Common::Io::Base;
+use pgBackRest::Common::Log;
+use pgBackRest::LibC qw(:cipher);
+use pgBackRest::Storage::Base;
+
+####################################################################################################################################
+# Package name constant
+####################################################################################################################################
+use constant STORAGE_FILTER_CIPHER_BLOCK                            => __PACKAGE__;
+    push @EXPORT, qw(STORAGE_FILTER_CIPHER_BLOCK);
+
+####################################################################################################################################
+# CONSTRUCTOR
+####################################################################################################################################
+sub new
+{
+    my $class = shift;
+
+    # Assign function parameters, defaults, and log debug info
+    my
+    (
+        $strOperation,
+        $oParent,
+        $strCipherType,
+        $tCipherKey,
+        $strMode,
+    ) =
+        logDebugParam
+        (
+            __PACKAGE__ . '->new', \@_,
+            {name => 'oParent', trace => true},
+            {name => 'strCipherType', trace => true},
+            {name => 'tCipherKey', trace => true},
+            {name => 'strMode', optional => true, default => STORAGE_ENCRYPT, trace => true},
+        );
+
+    # Bless with new class
+    my $self = $class->SUPER::new($oParent);
+    bless $self, $class;
+
+    # Check mode is valid
+    $self->{strMode} = $strMode;
```

Comment:
This is on purpose.  The variable that will actually be used later on is the one that is checked for validity so that the assignment is checked as well.  If an error occurs the value of the variable is not important.

Reply:
It just seems to me that it would be more efficient to check $strMode and then set $self->{strMode} = $strMode if $strMode it valid.

Reply:
I'm going for correctness here more than efficiency.  In all but the most extreme situations that assignment will run anyway.  The error is an ASSERT - so really it should never be hit at all.